fix: lay plumbing for having simulations throw an error if they cannot be added in a block

sklppy88 opened the pull request:
This PR gives an option to make a simulation of a tx fail, if the tx cannot be included in a block and added to the state.

To minimize impact on callsites / disruption, this update does not make this validation necessary, and defaults to skipping tx validation, exhibiting previous non-validated behavior. This will be changed to be default behavior in the future.

e.g. If a simulation produces duplicate nullifiers, or nullifiers that already exist in a state tree, the results of this simulation should not be returned, but should warn users that the transaction simulated is impossible to actually be added to a block due to being an invalid transaction.

The method for achieving the above is that a new API on the node was created, used for validating the correctness of the metadata and side-effects produced by a transaction. A transaction is deemed valid if and only if the transaction can be added to a block that can be used to advance state.

Another note just for reference: there is weirdness in e2e_prover, that fails the proof validation.

nventuro commented on yarn-project/aztec.js/src/contract/contract_function_interaction.ts:
```diff
@@ -23,6 +23,8 @@ export type SimulateMethodOptions = {
   from?: AztecAddress;
   /** Gas settings for the simulation. */
   gasSettings?: GasSettings;
+  /** Simulate without validating tx against current state. */
```
```suggestion
  /** Simulate without validating tx results against current state. */
```

nventuro commented on yarn-project/circuit-types/src/interfaces/aztec-node.ts:
```diff
@@ -316,6 +316,13 @@ export interface AztecNode {
   simulatePublicCalls(tx: Tx): Promise<PublicSimulationOutput>;
 
+  /**
+   * Validates the correctness of the execution, namely that a transaction is valid if and
+   * only if the transaction can be added to a valid block at the current state.
```
```suggestion
   * Returns true if the transaction is valid for inclusion at the current state. Valid transactions can be
   * made invalid by *other* transactions if e.g. they emit the same nullifiers, or come become invalid
   * due to e.g. the max_block_number property.
```

nventuro commented on yarn-project/pxe/src/pxe_service/pxe_service.ts:
```diff
@@ -497,6 +497,7 @@ export class PXEService implements PXE {
     txRequest: TxExecutionRequest,
     simulatePublic: boolean,
     msgSender: AztecAddress | undefined = undefined,
+    offline: boolean = true,
```
Why go with an offline default? To minimize disruption?

nventuro commented on yarn-project/aztec-node/src/aztec-node/server.ts:
```diff
@@ -339,9 +341,7 @@ export class AztecNodeService implements AztecNode {
-    const [_, invalidTxs] = await this.txValidator.validateTxs([tx]);
-    if (invalidTxs.length > 0) {
-      this.log.warn(`Rejecting tx ${tx.getTxHash()} because of validation errors`);
+    if ((await this.validateTx(tx)) === false) {
```
```suggestion
    if (! await this.validateTx(tx)) {
```

nventuro commented on yarn-project/circuit-types/src/interfaces/aztec-node.ts:
```diff
@@ -316,6 +316,13 @@ export interface AztecNode {
+   * only if the transaction can be added to a valid block at the current state.
+   * @param tx - The transaction to validate for correctness.
+   */
+  validateTx(tx: Tx): Promise<boolean>;
```
I'd consider renaming this to e.g. `isTxValid` - 'validate' to me hints at some side effect potentially taking place

nventuro commented on yarn-project/aztec-node/src/aztec-node/server.ts:
```diff
@@ -165,6 +166,7 @@ export class AztecNodeService implements AztecNode {
       new DataTxValidator(),
       new MetadataTxValidator(config.l1ChainId),
       new TxProofValidator(proofVerifier),
+      new DoubleSpendTxValidator(new WorldStateDB(worldStateSynchronizer.getLatest())),
```
Why is this validator being added here?

nventuro replied:
But isn't this PR about having `simulate` check with the node for note validity? What this change does is extend the validations the node performs to also check for double spends, which is unrelated.

nventuro commented:
Looks good! Included a minor naming suggestion, as I was left wondering for a bit what the method actually did. I'm now thinking too tha perhaps `offline` is not the best term, since we _are_ accesing the node to e.g. fetch public data, merkle paths, etc.

nventuro commented on yarn-project/aztec.js/src/contract/contract_function_interaction.ts:
```diff
@@ -23,6 +23,8 @@ export type SimulateMethodOptions = {
   from?: AztecAddress;
   /** Gas settings for the simulation. */
   gasSettings?: GasSettings;
+  /** Option to throw an error if simulation does not produce a valid transaction. */
+  ignoreInvalidTx?: boolean;
```
Hm better, but this makes me thing that an invalid tx would be ignored - which is actually the opposite of what this does, we 'ignore' the fact that it is invalid.

What about 'skipTxValidation'?

Commit:
init

fix: fixing private voting by correctly throwing an error if simulation fails (#7840)

This PR makes a simulation of a tx fail, if the tx cannot be included in
a block and added to the state.

e.g. If a simulation produces duplicate nullifiers, or nullifiers that
already exist in a state tree, the results of this simulation should not
be returned, but should warn users that the transaction simulated is
impossible to actually be added to a block due to being an invalid
transaction.

The method for achieving the above is that a new API on the node was
created, used for validating the correctness of the metadata and
side-effects produced by a transaction. A transaction is deemed valid if
and only if the transaction can be added to a block that can be used to
advance state.

Note: this update does not make this validation necessary, and defaults
to offline simulation. Offline simulation is previous non-validated
behavior, and is potentially useful if we ever move to a model where a
node is optional to a pxe.

Another note just for reference: there is weirdness in e2e_prover, that
fails the proof validation.

Resolves #4781.

Apply suggestions from code review

Co-authored-by: Nicolás Venturo <nicolas.venturo@gmail.com>

nventuro commented on yarn-project/aztec.js/src/contract/contract_function_interaction.ts:
```diff
@@ -23,6 +23,8 @@ export type SimulateMethodOptions = {
   from?: AztecAddress;
   /** Gas settings for the simulation. */
   gasSettings?: GasSettings;
+  /** Option to throw an error if simulation does not produce a valid transaction. */
```
```suggestion
  /** Simulate without checking for the validity of the resulting transaction, e.g. whether it emits any existing nullifiers. */
```

nventuro commented on yarn-project/pxe/src/pxe_service/pxe_service.ts:
```diff
@@ -513,6 +513,7 @@ export class PXEService implements PXE {
     txRequest: TxExecutionRequest,
     simulatePublic: boolean,
     msgSender: AztecAddress | undefined = undefined,
+    skipTxValidation: boolean = true,
```
```suggestion
    skipTxValidation: boolean = true, // TODO(#7956): make the default be false
```

nventuro reviewed (approved):
Looks good!

sklppy88 committed:
Apply Nico's suggestions from code review

Co-authored-by: Nicolás Venturo <nicolas.venturo@gmail.com>
